runtime/vm/extendedMessageNPE.cpp

```diff
@@ -22,10 +22,227 @@
 
 #include "cfreader.h"
 #include "stackwalk.h"
+#include "util_api.h"
 #include "ut_j9vm.h"
 
 extern "C" {
 
+static char* convertToJavaFullyQualifiedName(J9VMThread *vmThread, J9UTF8 *fullyQualifiedNameUTF);
+static char* convertMethodSignature(J9VMThread *vmThread, J9UTF8 *methodSig);
+
+
+/**
+ * Replace '/' with '.' for an internal fully qualified name
+ *
+ * The caller is responsible for freeing the returned string.
+ *
+ * @param[in] vmThread current J9VMThread
+ * @param[in] fullyQualifiedNameUTF pointer to J9UTF8 containing the fully qualified name
+ *
+ * @return a char pointer to the fully qualified name,
+ *         NULL if not successful, but keep application exception instead of throwing OOM.
+ */
+static char*
+convertToJavaFullyQualifiedName(J9VMThread *vmThread, J9UTF8 *fullyQualifiedNameUTF)
+{
+	PORT_ACCESS_FROM_VMC(vmThread);
+	UDATA length = J9UTF8_LENGTH(fullyQualifiedNameUTF);
+	char *result =  (char *)j9mem_allocate_memory(length + 1, OMRMEM_CATEGORY_VM);
+
+	if (NULL != result) {
+		char *cursor = result;
+		char *end = result + length;
+
+		memcpy(result, J9UTF8_DATA(fullyQualifiedNameUTF), length);
+		while (cursor < end) {
+			if ('/' == *cursor) {
+				*cursor = '.';
+			}
+			cursor += 1;
+		}
+		*end = '\0';
+	}
+
+	Trc_VM_ConvertToJavaFullyQualifiedName_Get_ClassName(vmThread, result, length, fullyQualifiedNameUTF);
+
+	return result;
+}
+
+/**
+ * Convert a signature string
+ * from
+ *  ([[[Ljava/lang/String;Ljava/lang/Void;BLjava/lang/String;ICDFJLjava/lang/Object;SZ)Ljava/lang/String;
+ * to
+ * (java.lang.String[][][], java.lang.Void, byte, java.lang.String, int, char, double, float, long, java.lang.Object, short, boolean)
+ *
+ * Note:
+ * The return type is ignored.
+ * Assuming the method signature is well formed.
+ * The caller is responsible for freeing the returned string.
+ *
+ * @param[in] vmThread current J9VMThread
+ * @param[in] methodSig pointer to J9UTF8 containing the method signature
+ *
+ * @return a char pointer to the class name within fullQualifiedUTF,
+ *         NULL if not successful, but keep application exception instead of OOM.
+ */
+static char*
+convertMethodSignature(J9VMThread *vmThread, J9UTF8 *methodSig)
+{
+	UDATA i = 0;
+	UDATA j = 0;
+	UDATA arity = 0;
+	U_8 *string = J9UTF8_DATA(methodSig);
+	UDATA stringLength = J9UTF8_LENGTH(methodSig);
+	UDATA bufferSize = 0;
+	char *result = NULL;
+
+	PORT_ACCESS_FROM_VMC(vmThread);
+
+	/* first scan to calculate the buffer size required */
+	/* first character is '(' */
+	bufferSize += 1;
+	for (i = 1; ((')' != string[i]) && (i < stringLength)); i++) {
+		arity = 0;
```

Comment:
Can arity be defined here in the loop?  And a new variable introduced in the other variable?  Neither lives past the end of the loop so better to encapsulate them in their own loops

Reply:
Yeah, moved the variable definition into the their own loops.